---
title: 스프링 MVC 2편 - 백엔드 웹 개발 활용 기술 - (43) 예외 처리와 오류 페이지
aliases: 
tags:
  - spring
  - exception
categories:
  - spring
toc: true
toc_label: 목차
date: 2024-06-19
last_modified_at: 2024-06-19
---
>  인프런 스프링 MVC 2편 - 백엔드 웹 개발 활용 기술편을 학습하고 정리한 내용 입니다.

예외를 어떻게 처리하면 좋을지 이번 시간부터 알아보자.
## 프로젝트 생성

[스프링 부트 스타터](https://start.spring.io/)이용

- 프로젝트 선택
	- Project : Gradle Project
	- Language : java
	- Spring boot : 3.3.0
- Project Metadata
	- Group : hello
	- Artifact : exception
	- Name : exception
	- Package name : **hello.exception**
	- Packaging : **Jar**
	- Java : 21
- Dependencies : **Spring Web, Lombok , Thymeleaf, Validation**

기본적으로 강의와 시간이 차이가 나서 스프링 부트와, 자바 버전이 좀 다르다.
심지어 올해 초와 지금 또 다르다.

![](https://i.imgur.com/uI1LUcI.png){: .align-center}


다음과 같이 세팅 한 후에 인텔리제이에서 띄워보자.

```
plugins {  
    id 'java'  
    id 'org.springframework.boot' version '3.3.0'  
    id 'io.spring.dependency-management' version '1.1.5'  
}  
  
group = 'hello'  
version = '0.0.1-SNAPSHOT'  
  
java {  
    toolchain {  
       languageVersion = JavaLanguageVersion.of(21)  
    }  
}  
  
```

그런데 또 build.gradle에서 `자바 부분`이 좀 다른 것 같다.

```
plugins {  
    id 'java'  
    id 'org.springframework.boot' version '3.2.5'  
    id 'io.spring.dependency-management' version '1.1.4'  
}  
  
group = 'hello'  
version = '0.0.1-SNAPSHOT'  
  
java {  
    sourceCompatibility = '21'  
}
```

부트 3.2.x 일 때는 다음과 같았다..


![](https://i.imgur.com/rAA18Zx.png){: .align-center}

다음과 같이 뜨면 성공이다.


## 서블릿 예외 처리 - 시작

스프링이 아닌 순수 서블릿 컨테이너는 예외를 어떻게 처리하는지 알아보자.

> 서블릿은 다음 2가지 방식으로 예외 처리를 지원한다.
- `Exception`(예외)
- `response.sendError(HTTP 상태 코드, 오류 메시지)`

### Exception(예외)

**자바 직접 실행**

자바의 메인 메서드를 직접 실행하는 경우 `main`이라는 이름의 쓰레드가 실행된다.

실행 도중에 예외를 잡지 못하고 처음 실행한 `main()`메서드를 넘어서 예외가 던져지면, 예외 정보를 남기고 해당 쓰레드는 종료 된다.

**웹 애플리케이션**

웹 애플리케이션은 사용자 요청 별로 별도의 **쓰레드**가 할당되고, 서블릿 컨테이너 안에서 실행된다.

애플리케이션에서 예외가 발생했는데, 어디서 try ~ catch로 예외를 잡아서 처리하면 아무 문제가 없다.

그런데 애플리케이션에서 예외를 잡지 못하고, 서블릿 밖으로 까지 예외가 전달되면 어떻게 되나?

```
WAS(여기 까지 전파) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(예외 발생)
```

결국 톰캣 같은 WAS까지 예외가 전달된다. WAS는 예외가 발생하면 어떻게 처리할까?

정확한 테스트를 위해 스프링 부트가 제공하는 기본 예외 페이지를 꺼두자.


**application.yml**
```yml
server:  
  error:  
    whitelabel:  
      enabled: false
```

간단하게 컨트롤러 하나 만들어서 예외를 던져 보자.

`hello.exception.servlet.ServletExController`
```java
@Slf4j  
@Controller  
public class ServletExController {  
  
    @GetMapping("/error-ex")  
    public String errorEx() {  
        throw new RuntimeException("예외 발생!!");  
    }  
}
```

실행해서 `/error-ex`로 들어가 보자.

![](https://i.imgur.com/CEL1ec1.png){: .align-center}

HTTP 상태 코드가 500으로 나온다.

`Exception`의 경우 서버 내부에서 처리할 수 없는 오류가 발생한 것으로 생각해서 HTTP 상태 코드 500을 반환한다.

이번에는 없는 URL을 한번 들어가 보자.

![](https://i.imgur.com/UM49zkW.png){: .align-center}

404 상태 코드가 나오고 톰캣이 기본 제공하는 화면이 나온다.

### response.sendError(HTTP 상태 코드, 오류 메시지)

오류가 발생했을 때 `HttpServletResponse`가 제공하는 `sendError`라는 메서드를 사용해도 된다.

이것을 호출한다고 당장 예외가 발생하는 것은 아니지만, 서블릿 컨테이너에게 오류가 발생했다는 점을 전달할 수 있다.

이 메서드를 사용하면 HTTP 상태 코드와 오류 메시지도 추가할 수 있다.

- `response.sendError(HTTP 상태 코드)`
- `response.sendError(HTTP 상태 코드, 오류 메시지)`

컨트롤러에 추가해 보자.


**ServletExController - 추가**
```java
@GetMapping("/error-404")  
public void error404(HttpServletResponse response) throws IOException {  
    response.sendError(404, "404 오류 !!!");  
}  
  
@GetMapping("/error-500")  
public void error500(HttpServletResponse response) throws IOException {  
    response.sendError(500);  
}
```

**sendError 흐름**

```
WAS(sendError 호출 기록 확인) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(response.sendError())
```

`response.sendError()`를 호출하면 `response` 내부에는 오류가 발생했다는 상태를 저장해둔다.

그리고 서블릿 컨테이너는 고객에게 응답 전에 `response`에 `sendError()`가 호출 되었는지 확인한다.

그리고 호출 되었다면 설정한 오류 코드에 맞추어 기본 오류 페이지를 보여준다.

![](https://i.imgur.com/UACxtKK.png){: .align-center}

![](https://i.imgur.com/fqLPx97.png){: .align-center}


이제 예외를 보내는 방법은 알았으니 화면을 제공해 보자.

## 서블릿 예외 처리 - 오류 화면 제공

서블릿은 `Exception`이 발생해서 서블릿 밖으로 전달이 되거나 또는 `response.sendError()`가 호출 되었을 때 각각의 상황에 맞는 오류 처리 기능을 제공한다.

과거에는 `web.xml`이라는 파일에 다음과 같이 오류 화면을 등록했다.

![](https://i.imgur.com/2heBR8A.png){: .align-center}

지금은 스프링 부트를 통해서 서블릿 컨테이너를 실행하기 때문에, 스프링 부트가 제공하는 기능을 사용해서 서블릿 오류 페이지를 등록하면 된다.


### 서블릿 오류 페이지 등록

![](https://i.imgur.com/fOTRmPF.png){: .align-center}

다음과 같이 `hello.exception.WebServerCustomizer`를 만들었다.

```java
@Component  
public class WebServerCustomizer implements WebServerFactoryCustomizer<ConfigurableServletWebServerFactory> {  
  
    @Override  
    public void customize(ConfigurableServletWebServerFactory factory) {  
        ErrorPage errorPage404 = new ErrorPage(HttpStatus.NOT_FOUND, "/error-page/404");  
        ErrorPage errorPage500 = new ErrorPage(HttpStatus.INTERNAL_SERVER_ERROR, "/error-page/500");  
  
        ErrorPage errorPageEx = new ErrorPage(RuntimeException.class, "/error-page/500");  
  
        factory.addErrorPages(errorPage404, errorPage500, errorPageEx);  
    }  
}
```

- `implements WebServerFactoryCustomizer<ConfigurableServletWebServerFactory>` 이거 구현해야 함.
- `response.sendError(404)` : `errorPage404` 호출
- `response.sendError(500)` : `errorPage500` 호출
- `RuntimeException` 또는 그 자식 타입의 예외 : `errorPageEx` 호출

500 예외가 서버 내부에서 발생한 오류라는 뜻을 포함하고 있기 때문에 여기 서는 예외가 발생한 경우도 500 오류 화면으로 처리했다.

오류 페이지는 예외를 다룰 때 해당 예외와 그 자식 타입의 오류를 함께 처리한다. 예를 들어서 위의 경우 `RuntimeException`은 물론이고 `RuntimeException`의 자식도 함께 처리한다.

오류가 발생했을 때 처리할 수 있는 컨트롤러가 필요하다.

`hello.exception.servlet.ErrorPageController`
```java
@Slf4j  
@Controller  
public class ErrorPageController {  
  
    @RequestMapping("/error-page/404")  
    public String errorPage404(HttpServletRequest request, HttpServletResponse response) {  
        log.info("error-page 404");  
        return "error-page/404";  
    }  
  
    @RequestMapping("/error-page/500")  
    public String errorPage500(HttpServletRequest request, HttpServletResponse response) {  
        log.info("error-page 500");  
        return "error-page/500";  
    }  
}
```

다음과 같이 컨트롤러를 만들었으니 이제 뷰 템플릿을 만들자.

![](https://i.imgur.com/Wky8j0E.png){: .align-center}


```html
<!DOCTYPE HTML>  
<html xmlns:th="http://www.thymeleaf.org">  
<head>  
    <meta charset="utf-8">  
</head>  
<body>  
  
<div class="container" style="max-width: 600px">  
    <div class="py-5 text-center">  
        <h2>404 오류 화면</h2>  
    </div>  
  
    <div>  
        <p>오류 화면 입니다.</p>  
    </div>  
  
    <hr class="my-4">  
  
</div> <!-- /container -->  
  
</body>  
</html>
```

```html
<!DOCTYPE HTML>  
<html xmlns:th="http://www.thymeleaf.org">  
<head>  
    <meta charset="utf-8">  
</head>  
<body>  
  
<div class="container" style="max-width: 600px">  
    <div class="py-5 text-center">  
        <h2>500 오류 화면</h2>  
    </div>  
  
    <div>  
        <p>오류 화면 입니다.</p>  
    </div>  
  
    <hr class="my-4">  
  
</div> <!-- /container -->  
  
</body>  
</html>
```

![](https://i.imgur.com/y3xgFOU.png){: .align-center}

500에러를 발생 시키는 컨트롤러를 호출 하니 500페이지로 이동했다.

![](https://i.imgur.com/oZXhj75.png){: .align-center}

세팅해둔 404 에러 페이지도 잘 동작한다.

## 서블릿 예외 처리 - 오류 페이지 작동 원리

서블릿은 `Exception`(예외)가 발생해서 서블릿 밖으로 전달되거나 또는 `response.sendError()`가 호출 되었을 때 설정된 오류 페이지를 찾는다.

**예외 발생 흐름**
```
WAS(여기 까지 전파) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(예외 발생)
```

**sendError 흐름**
```
WAS(sendError 호출 기록 확인) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(response.sendError())
```

WAS는 해당 예외를 처리하는 오류 페이지 정보를 확인한다.

`new ErrorPage(RuntimeException.class, "/error-page/500")`

예를 들어서 `RuntimeException`이 WAS까지 전달되면, WAS는 오류 페이지 정보를 확인한다.

확인해보니 `RuntimeException`은 `/error-page/500`으로 지정되어 있다.

WAS는 오류 페이지 출력을 위해 다시 `/error-page/500`을 요청한다.

**오류 페이지 요청 흐름**

```
WAS '/error-page/500' 다시 요청 -> 필터 -> 서블릿 -> 인터셉터 -> 컨트롤러(/error-page/500) -> View
```

**예외 발생과 오류 페이지 요청 흐름**

```
1. WAS(여기 까지 전파) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(예외 발생)

2. WAS '/error-page/500' 다시 요청 -> 필터 -> 서블릿 -> 인터셉터 -> 컨트롤러(/error-page/500) -> View
```


**중요한 점은 웹 브라우저(클라이언트)는 서버 내부에서 이런 일이 일어나는지 전혀 모른다는 점이다. 오직 서버 내부에서 오류 페이지를 찾기 위해 추가적인 호출을 한다.**

정리하면 다음과 같다.
1. 예외가 발생해서 WAS까지 전파된다.
2. WAS는 오류 페이지 경로를 찾아서 내부에서 오류 페이지를 호출한다. **이때 오류 페이지 경로로 필터, 서블릿, 인터셉터, 컨트롤러가 모두 다시 호출된다.**


### 오류 정보 추가

WAS는 오류 페이지를 단순히 다시 요청만 하는 것이 아니라, 오류 정보를 `request`의 `attribute`에 추가해서 넘겨준다.

필요하면 오류 페이지에서 이렇게 전달된 오류 정보를 사용할 수 있다.

**ErrorPageController - 오류 출력**

```java
@Slf4j  
@Controller  
public class ErrorPageController {  
  
 
    //RequestDispatcher 상수로 정의되어 있음  
	public static final String ERROR_EXCEPTION = "jakarta.servlet.error.exception";  
	public static final String ERROR_EXCEPTION_TYPE = "jakarta.servlet.error.exception_type";  
	public static final String ERROR_MESSAGE = "jakarta.servlet.error.message";  
	public static final String ERROR_REQUEST_URI = "jakarta.servlet.error.request_uri";  
	public static final String ERROR_SERVLET_NAME = "jakarta.servlet.error.servlet_name";  
	public static final String ERROR_STATUS_CODE = "jakarta.servlet.error.status_code";  
  
    @RequestMapping("/error-page/404")  
    public String errorPage404(HttpServletRequest request, HttpServletResponse response) {  
        log.info("error-page 404");  
        printErrorInfo(request);  
        return "error-page/404";  
    }  
  
    @RequestMapping("/error-page/500")  
    public String errorPage500(HttpServletRequest request, HttpServletResponse response) {  
        log.info("error-page 500");  
        printErrorInfo(request);  
        return "error-page/500";  
    }  
  
    private void printErrorInfo(HttpServletRequest request) {  
        log.info("ERROR_EXCEPTION: {}", request.getAttribute(ERROR_EXCEPTION));  
        log.info("ERROR_EXCEPTION_TYPE: {}", request.getAttribute(ERROR_EXCEPTION_TYPE));  
        log.info("ERROR_MESSAGE: {}", request.getAttribute(ERROR_MESSAGE));  
        log.info("ERROR_REQUEST_URI: {}", request.getAttribute(ERROR_REQUEST_URI));  
        log.info("ERROR_SERVLET_NAME: {}", request.getAttribute(ERROR_SERVLET_NAME));  
        log.info("ERROR_STATUS_CODE: {}", request.getAttribute(ERROR_STATUS_CODE));  
        log.info("dispatchType={}", request.getDispatcherType());  
    }  
}
```

다음과 같이 

![](https://i.imgur.com/8ow9FNM.png){: .align-center}


RequestDispatcher에 선언되어 있는 예외 정보들 중 몇 개를 가져와서 지금 어떤 예외가 터진 건지 로그로 확인해 보자.

```java
private void printErrorInfo(HttpServletRequest request) {  
	log.info("ERROR_EXCEPTION: {}", request.getAttribute(ERROR_EXCEPTION));  
	log.info("ERROR_EXCEPTION_TYPE: {}", request.getAttribute(ERROR_EXCEPTION_TYPE));  
	log.info("ERROR_MESSAGE: {}", request.getAttribute(ERROR_MESSAGE));  
	log.info("ERROR_REQUEST_URI: {}", request.getAttribute(ERROR_REQUEST_URI));  
	log.info("ERROR_SERVLET_NAME: {}", request.getAttribute(ERROR_SERVLET_NAME));  
	log.info("ERROR_STATUS_CODE: {}", request.getAttribute(ERROR_STATUS_CODE));  
	log.info("dispatchType={}", request.getDispatcherType());  
}  
```

request에 담긴 예외 정보를 출력할 것이다.

404 에러 

![](https://i.imgur.com/yJi3wnL.png){: 
.align-center}

500 에러

![](https://i.imgur.com/99pi5wV.png){: .align-center}


**request.attribute에 서버가 담아준 정보**
- `jakarta.servlet.error.exception` : 예외 
- `jakarta.servlet.error.exception_type` : 예외 타입
- `jakarta.servlet.error.message` : 오류 메시지 
- `jakarta.servlet.error.request_uri` : 클라이언트 요청 URI 
- `jakarta.servlet.error.servlet_name` : 오류가 발생한 서블릿 이름 
- `jakarta.servlet.error.status_code` : HTTP 상태 코드


> 참고 : 강의에선 `javax.servlet..` 이지만 자바 버전 때문에 `jakarta.servlet..` 으로 해줘야 함.